presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java

```diff
@@ -438,4 +439,37 @@ else if (operand >= upper) {
 
         return result;
     }
+
+    @Description("The bucket number of a value given an array of bins")
+    @ScalarFunction("width_bucket")
+    @SqlType(StandardTypes.BIGINT)
+    public static long widthBucket(@SqlType(StandardTypes.DOUBLE) double operand, @SqlType("array(double)") Block bins)
+    {
+        int nbins = bins.getPositionCount();
+
+        checkCondition(nbins > 0, INVALID_FUNCTION_ARGUMENT, "Bins cannot be an empty array");
+        checkCondition(!Double.isNaN(operand), INVALID_FUNCTION_ARGUMENT, "Operand cannot be NaN");
+
+        int lower = 0;
+        int upper = nbins;
+
+        int index;
+        double bin;
+
+        while (lower < upper) {
```

Comment:
We can't cover all cases here without traversing the entire list, which would defeat the purpose of using binary search here. For example,

```sql
select width_bucket(1.5, array[1.0, 2.3, 2.0])
```

Let's say we add a check that `bins[lower] > bins[upper - 1]`.

It would operate like this:

1st iteration:
1. check that `bins[lower] > bins[upper - 1]`, this is `false` so we continue on
1. `index = (0 + 3) / 2 => index == 1`
1. `bin = bins[index] => bin == 2.3`
1. `operand < bin => upper == index == 1`

So, we'll never reach the `2.0` value.

Should we still add this check? It would catch less ambiguous cases like bins in sorted descending order, but it won't catch many things that have sorted run in them.

Personally, I don't think this is a huge issue, as I imagine most uses of this function will be with an array constant.